src/core/eggmath.rkt

```diff
@@ -0,0 +1,213 @@
+  (cond
+    [(list? expr)
+     (string-append
+      "("
+      (symbol->string (first expr))
+      (foldr
+       (lambda (sub-expr acc)
+         (string-append " " (expr->egg-expr-helper sub-expr egg->herbie-dict herbie->egg-dict)
+                        acc))
+       "" (rest expr))
+      ")")]
+    [(and (number? expr) (exact? expr) (real? expr))
+     (number->string expr)]
+    [(hash-has-key? herbie->egg-dict expr)
+     (symbol->string (hash-ref herbie->egg-dict expr))]
+    [else
+     (define new-key (string-append "h" (number->string (hash-count herbie->egg-dict))))
+     (define new-key-symbol (string->symbol new-key))
+         
+     (hash-set! herbie->egg-dict
+                expr
+                new-key-symbol)
+     (hash-set! egg->herbie-dict
+                new-key-symbol
+                expr)
+      new-key]))
+
+
+(define-struct (egg-add-exn exn:fail:user) ())
+
+;; result function is a function that takes the ids of the nodes
+;; egraph-add-exprs returns the result of result-function
+(define (egraph-add-exprs eg-data exprs result-function)
+  (define egg-exprs
+    (for/list ([expr exprs])
+      (expr->egg-expr expr eg-data)))
+  (debug #:from 'simplify (format "Sending expressions to egg_math:\n ~a"
+                                  (string-join (map ~a egg-exprs) "\n ")))
+      
+  (define expr-results
+    (for/list ([expr egg-exprs])
+      (egraph_add_expr (egraph-data-egraph-pointer eg-data) expr)))
+  
+  (define node-ids
+    (for/list ([result expr-results])
+      (if (EGraphAddResult-successp result)
+          (EGraphAddResult-id result)
+          (raise (egg-add-exn
+               (string-append "Failed to add expr to egraph")
+               (current-continuation-marks))))))
+
+  (define res (result-function node-ids))
+
+  (for/list ([result expr-results])
+    (egraph_addresult_destroy result))
+
+  res)
+
+
+(module+ test
+  (require (submod "../syntax/rules.rkt" internals))
+  
+  (define all-simplify-rules
+    (for/append ([rec (*rulesets*)])
+      (match-define (list rules groups _) rec)
+      (if 
+       (set-member? groups 'simplify)
+       rules
+       '())))
+
+  ;; check that no rules in simplify match on bare variables
+  ;; this would be bad because we don't want to match type-specific operators on a value of a different type
+  (for ([rule all-simplify-rules])
+    (check-true
+     (or
+      (not (symbol? (rule-input rule)))
+      (set-member? fpconstants (rule-input rule)))
+     (string-append "Rule failed: " (symbol->string (rule-name rule)))))
+
+  
+  (check-equal?
+   (egraph-run
+    (lambda (egg-graph)
+      (egraph-add-exprs
+       egg-graph
+       (list '(+ x 2) '(+ x 1) '(+ x 0) '(+ x 23) 0 1)
+       (lambda (node-ids)
+         (egraph-run-rules egg-graph 100 (*simplify-rules*) true)
+         (for/list [(node-id node-ids)]
+           (egg-expr->expr (egraph-get-simplest egg-graph node-id) egg-graph))))))
```

Comment:
Is this going to be a problem if the e-graph commutes the additions? And is there any reason not to just have these as part of the broader simplify tests? It seems like we duplicate a lot.

Reply:
If it commutes additions, it won't pick the new one because it won't have a lower cost.
I like this test more as a test that the functions all work. It's a good example of how to use `eggmath.rkt` as well, since it can be confusing.

Reply:
Perhaps that's reason to move it into the `herbie-egg` package and replace the reference to `*simplify-rules*` with some simple, hard-coded list of rules. Anyway, sounds good.

Reply:
On second thought, I'm removing this test for now. It doesn't belong in here.